pybamm/expression_tree/array.py

```diff
@@ -57,6 +57,30 @@ def __init__(
             name, domain=domain, auxiliary_domains=auxiliary_domains, domains=domains
         )
 
+    @classmethod
+    def _from_json(cls, snippet: dict):
```

Comment:
Given that PyBaMM supports Python 3.8+, each file using the newer type hints should import -
```py
from __future__ import annotations
```
at the top. This can also be automated using the `isort` rules in `ruff`.